server/src/main/java/org/opensearch/indices/store/TransportNodesListShardStoreMetadataBatch.java

```diff
@@ -0,0 +1,525 @@
+
+        /**
+         * @return commit sync id if exists, else null
+         */
+        public String syncId() {
+            return metadataSnapshot.getSyncId();
+        }
+
+        @Override
+        public String toString() {
+            return "StoreFilesMetadata{"
+                + ", shardId="
+                + shardId
+                + ", metadataSnapshot{size="
+                + metadataSnapshot.size()
+                + ", syncId="
+                + metadataSnapshot.getSyncId()
+                + "}"
+                + '}';
+        }
+    }
+
+    /**
+     * Request is used in constructing the request for making the transport request to set of other node.
+     * Refer {@link TransportNodesAction} class start method.
+     *
+     * @opensearch.internal
+     */
+    public static class Request extends BaseNodesRequest<Request> {
+
+        private final Map<ShardId, String> shardIdsWithCustomDataPath;
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            shardIdsWithCustomDataPath = in.readMap(ShardId::new, StreamInput::readString);
+        }
+
+        public Request(Map<ShardId, String> shardIdsWithCustomDataPath, DiscoveryNode[] nodes) {
+            super(nodes);
+            this.shardIdsWithCustomDataPath = Objects.requireNonNull(shardIdsWithCustomDataPath);
+        }
+
+        public Map<ShardId, String> getShardIdsWithCustomDataPath() {
+            return shardIdsWithCustomDataPath;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeMap(shardIdsWithCustomDataPath, (o, k) -> k.writeTo(o), StreamOutput::writeString);
+        }
+    }
+
+    /**
+     * Metadata for the nodes store files
+     *
+     * @opensearch.internal
+     */
+    public static class NodesStoreFilesMetadataBatch extends BaseNodesResponse<NodeStoreFilesMetadataBatch> {
+
+        public NodesStoreFilesMetadataBatch(StreamInput in) throws IOException {
+            super(in);
+        }
+
+        public NodesStoreFilesMetadataBatch(
+            ClusterName clusterName,
+            List<NodeStoreFilesMetadataBatch> nodes,
+            List<FailedNodeException> failures
+        ) {
+            super(clusterName, nodes, failures);
+        }
+
+        @Override
+        protected List<NodeStoreFilesMetadataBatch> readNodesFrom(StreamInput in) throws IOException {
+            return in.readList(NodeStoreFilesMetadataBatch::new);
+        }
+
+        @Override
+        protected void writeNodesTo(StreamOutput out, List<NodeStoreFilesMetadataBatch> nodes) throws IOException {
+            out.writeList(nodes);
+        }
+    }
+
+    /**
+     * The metadata for the node store files
+     *
+     * @opensearch.internal
+     */
+    public static class NodeStoreFilesMetadata {
+
+        private StoreFilesMetadata storeFilesMetadata;
+        private Exception storeFileFetchException;
```

Comment:
Earlier for StartedShards(PSA's transport action) we were storing per shard exception in the response. But for StoreShards(RSA's transport action) we were were directly throwing exception. With your changes now we will be storing exception per shard. Can you please verify what will be the implication of it in processing in master [here](https://github.com/opensearch-project/OpenSearch/blob/2.3/server/src/main/java/org/opensearch/gateway/AsyncShardFetch.java#L200)

Scenario:
1. Just throw exception from data node over transport, and verify it is processed in List<T> responses rather than in failures condition. If this is true that you will not be changing any failure handling at master.

Reply:
[I induced the error on the data node](https://github.com/Gaurav614/OpenSearch/commit/78794d7ce5997bfe960499e4af659c81de7847ef#diff-ddb9033a1ecb36461c97bd2f4e06e8a18946c965f62ae4b3bdd15046edf8590a), it is processed in the onResponse as expected instead of onFailure case. It needs to be handled [like this](https://github.com/Gaurav614/OpenSearch/blob/async-shard-fetch-replica-induced-error/server/src/main/java/org/opensearch/gateway/GatewayAllocator.java#L526), I think this will be a separate task.

Log on Data 1 node
```
[2023-07-20T23:52:01,055][INFO ][o.o.i.s.TransportNodesListShardStoreMetadataBatch] [data1] sdarb Inducing error on node data1
```

Log on  Master node Gatewayallocator
```
[2023-07-20T23:53:08,826][ERROR][o.o.g.G.TestInternalReplicaShardAllocator] [master1] sdarb There was an error while fetching replica store data for shard [my-index-4][0], node[DQIxpsnNRA24tv_C4tkNOQ], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=H9uklZEHS3iUTPAnn6b86g], unassigned_info[[reason=NODE_LEFT], at[2023-07-20T18:20:23.323Z], delayed=false, details[node_left [Phd2Sj8gT26Kc3OvIxtm0w]], allocation_status[no_attempt]] on node {data1}{Phd2Sj8gT26Kc3OvIxtm0w}{l0V0RsyIRL62yNcYCbjudA}{127.0.0.1}{127.0.0.1:9302}{dir}{shard_indexing_pressure_enabled=true} so skipping this store data
```
Log on Data 1 node Master node AsyncShardFetch
```
[2023-07-20T23:53:07,958][INFO ][o.o.g.GatewayAllocator   ] [master1] Got the response
```

Reply:
@sudarshan-baliga  Thanks. 
@shiv0408 Please make a note of while making changes for ReplicaShardAllocator